internal/registry/registry_cache_test.go

```diff
@@ -0,0 +1,227 @@
+			h.AssertEq(t, bp.Version, "1.2.0")
+		})
+
+		it("locates a buildpack with version", func() {
+			bp, err := registryCache.LocateBuildpack("example/foo@1.1.0")
+			h.AssertNil(t, err)
+			h.AssertNotNil(t, bp)
+
+			h.AssertEq(t, bp.Namespace, "example")
+			h.AssertEq(t, bp.Name, "foo")
+			h.AssertEq(t, bp.Version, "1.1.0")
+		})
+
+		it("returns error if can't parse buildpack id", func() {
+			_, err := registryCache.LocateBuildpack("quack")
+			h.AssertError(t, err, "parsing buildpacks registry id")
+		})
+
+		it("returns error if buildpack id is empty", func() {
+			_, err := registryCache.LocateBuildpack("example/")
+			h.AssertError(t, err, "empty buildpack name")
+		})
+
+		it("returns error if can't find buildpack with requested id", func() {
+			_, err := registryCache.LocateBuildpack("example/qu")
+			h.AssertError(t, err, "reading entry")
+		})
+
+		it("returns error if can't find buildpack with requested version", func() {
+			_, err := registryCache.LocateBuildpack("example/foo@3.5.6")
+			h.AssertError(t, err, "could not find version")
+		})
+	})
+
+	when("#Refresh", func() {
+		var (
+			registryCache Cache
+		)
+
+		it.Before(func() {
+			registryCache, err = NewRegistryCache(logger, tmpDir, registryFixture)
+			h.AssertNil(t, err)
+		})
+
+		when("registry has new commits", func() {
+			it.Before(func() {
+				err := registryCache.Refresh()
+				h.AssertNil(t, err)
+
+				h.AssertGitHeadEq(t, registryFixture, registryCache.Root)
+
+				r, err := git.PlainOpen(registryFixture)
+				h.AssertNil(t, err)
+
+				w, err := r.Worktree()
+				h.AssertNil(t, err)
+
+				commit, err := w.Commit("second", &git.CommitOptions{
+					Author: &object.Signature{
+						Name:  "John Doe",
+						Email: "john@doe.org",
+						When:  time.Now(),
+					},
+				})
+				h.AssertNil(t, err)
+
+				_, err = r.CommitObject(commit)
+				h.AssertNil(t, err)
+			})
+
+			it("pulls the latest index", func() {
+				h.AssertNil(t, registryCache.Refresh())
+				h.AssertGitHeadEq(t, registryFixture, registryCache.Root)
+			})
+		})
+
+		when("Root is an empty string", func() {
+			it("fails to refresh", func() {
+				registryCache.Root = ""
+				err = registryCache.Refresh()
+				h.AssertError(t, err, "could not initialize")
+			})
+		})
+	})
+
+	when("#Initialize", func() {
+		var (
+			registryCache Cache
+		)
+
+		it.Before(func() {
+			registryCache, err = NewRegistryCache(logger, tmpDir, registryFixture)
+			h.AssertNil(t, err)
+		})
+
+		when("root is empty string", func() {
+			it.Before(func() {
+				registryCache.Root = ""
+			})
+
+			it("fails to create registry cache", func() {
+				err = registryCache.Initialize()
+				h.AssertError(t, err, "could not create registry cache")
+			})
+
+			when("url is empty string", func() {
+				it("fails to clone cache", func() {
+					registryCache.url = &url.URL{}
+
+					err = registryCache.Initialize()
+					h.AssertError(t, err, "could not clone remote registry")
+				})
+			})
+		})
```

Comment:
There's quite a lot of nesting happening here, none of these cases have any other state than `registryCache.Root = ""` so could that be moved in to the parent's `Before` (or is it even the default returned by `NewRegistryCache`?)

Does "url is an empty string" need to be a sub-context of "root is empty string"? On the surface they don't seem to be related. My guess is the check for an empty URL occurs before the check for an empty root so these tests will pass. Does that order need to be guaranteed in this way?

Reply:
These cases were contrived to try and force some of the error cases, and would definitely not be standard. The nesting of url and root both being empty was actually relevant, because when the Root is empty, it mostly passes (it fetches the cache to a new `Tmpdir`), it just fails on the `os.Rename` at the end of the `createCache` func (although, thinking about it, we could easily make it more resilient and not fail in that case, as unlikely as it is). However, when the `url` is empty, then it fails to even fetch the cache in the first place.